src/pip/_internal/utils/urls.py

```diff
@@ -13,13 +14,22 @@ def get_url_scheme(url: str) -> Optional[str]:
     return url.split(":", 1)[0].lower()
 
 
+@lru_cache(maxsize=None)
```

Comment:
This is unbounded and would store information that is not useful.

Reply:
Yes, but what other solution is there?

We can't know ahead of time how many file paths need caching. 

If a maxsize is given it is completely arbitrary. If you think it's required for memory safety I would prefer a very large number that is unexpected to be reached, like 10'000.

Reply:
> Yes, but what other solution is there?

As you said above

> An alternative solution would be to rearchitect Pip to not need to call path_to_url so much

Agreed, it's complex, but that may be better than throwing memory at the problem - after all, pip does get used in memory-constrained environments. I don't know what `lru_cache` does when it's getting close to memory limits, but I doubt it tries to manage that situation particularly - so you'd probably at some point start to get paging and a significant *reduction* in performance.

> We can't know ahead of time how many file paths need caching.

No, but it's not a matter of *needing* to cache anything. It's simply a case of only getting some of the performance benefits, not all of them.

Ultimately, we need to balance different use cases here. I'd consider installing 1000 wheels in a single install to be a very extreme case, and honestly I don't think 2.5 minutes is a particularly bad time for that. So while I'm always glad if we can get performance improvements, I think we have to be careful to keep perspective here. The overwhelming majority of the uses of pip "in the wild" are likely to be of the form `pip install <one or two packages>`.

Reply:
Alternatively, maybe we can just speed up `path_to_url`? Looking at the following:

```
❯ pyperf timeit -s "from pathlib import Path; import os" "Path(os.path.normpath(os.path.abspath('.'))).as_uri()"
.....................
Mean +- std dev: 5.67 us +- 0.13 us
❯ pyperf timeit -s "from urllib.parse import urljoin; from urllib.request import pathname2url; import os" "urljoin('file:', pathname2url(os.path.normpath(os.path.abspath('.'))))"
.....................
Mean +- std dev: 9.23 us +- 0.34 us
```

suggests that using `Path.as_uri()` is a lot faster. Using `Path.absolute()` rather than `os.path` loses a lot of the gain, I'm not sure why - maybe because it calls the `Path` constructor twice.

The point is, there may well be other options than caching the results. Or there may be improvements that can be achieved *as well as* a (limited-size) cache. As with any performance exercise, it's all about trade offs.

Reply:
> I'd consider installing 1000 wheels in a single install to be a very extreme case, and honestly I don't think 2.5 minutes is a particularly bad time for that. 

Home assistant is a very popular application in the smart home world, so it's relatively common use case.

And it's 2.5 minutes in my machine, it's 1-2 hours on others people's: https://github.com/pypa/pip/issues/12314. This was going to my first in a series of PRs.

And the cache here only grows if there are a lot of wheels, so there's only a few kilobytes of memory used in non-"extreme" examples.

> Agreed, it's complex, but that may be better than throwing memory at the problem - after all, pip does get used in memory-constrained environments. 

If a user needs to install over 1000 wheels using Pip I have to assume they have 1 or 2 MBs of spare memory.

I'll take another look at rearchitecture approach, but it probably means a significant rework of the way pip and resolvelib interact with each other. And my worry is that even I was able to make a PR there's a good chance it would never be accepted as a non-Pip maintainer adding such a significant requirement of knowledge for maintenance.

Reply:
Also, `os.path.abspath` includes a `normpath` call, so that's redundant if you're looking for performance improvements...

Reply:
> Home assistant is a very popular application in the smart home world, so it's relatively common use case.

I don't want to make this about one application, but common relative to what? I'll repeat, I'm not against performance improvements, but I'd like to understand the broader impact here.

> I'll take another look at rearchitecture approach, but it probably means a significant rework of the way pip and resolvelib interact with each other.

Why? All we're talking about is another form of canonicalisation, very similar to canonicalising project names or extras. I don't see how it would impact the resolution process at all.

Maybe we're talking at cross purposes. When I'm thinking about rearchitecting, I mean that we calculate the URL once for each object, and store it on that object. In effect replacing the cache you're suggesting with a precalculated attribute. The advantage is that the URLs get discarded when the objects get discarded, rather than being kept around forever. I don't know how important that is (I may not be as concerned with unbounded cache growth as @pradyunsg is) but it's a *lot* easier to reason about if it's an object attribute.

Reply:
> Also, os.path.abspath includes a normpath call, so that's redundant if you're looking for performance improvements...

I was thinking about that but reading the documention for [abspath](https://docs.python.org/3/library/os.path.html#os.path.abspath):

> Return a normalized absolutized version of the pathname path. On most platforms, this is equivalent to calling the function [normpath()](https://docs.python.org/3/library/os.path.html#os.path.normpath) as follows: normpath(join(os.getcwd(), path)).

I did not like the look of "On _most_ platforms", and my assumption what the reason it was both made absolute and normal was because of a motivating issue on some platform Pip is used on.

> I don't want to make this about one application, but common relative to what? I'll repeat, I'm not against performance improvements, but I'd like to understand the broader impact here.

Well, I guess I mean common relative to one person posting an issue on Pip for their personal project, and that at least 1000s of people have this use case because of the popularity of the project implementing this approach.

But moreover this is the general trend, installations involving an increasingly large number of wheels. In part because of Pip's resolver, implemented in 2020, that is supposed to guarantee correct resolution at install time there is greater safety in having a deeper and wider stack of dependencies to resolve. And in part because wheels don't require assumptions about a users environments like sdists do. And in part because we are told wheels are better: https://pradyunsg.me/blog/2022/12/31/wheels-are-faster-pure-python/

> Why? All we're talking about is another form of canonicalisation, very similar to canonicalising project names or extras. I don't see how it would impact the resolution process at all.

I could be very wrong here, mostly I was looking at the call graph and seeing near the top the motivating calls are coming from resolvelib.

> Maybe we're talking at cross purposes. When I'm thinking about rearchitecting, I mean that we calculate the URL once for each object, and store it on that object. In effect replacing the cache you're suggesting with a precalculated attribute. 

I certainly was not thinking like that, I think if that's possible it will need to be done around here: https://github.com/pypa/pip/blob/23.2.1/src/pip/_internal/index/package_finder.py#L841. That's where the number of calls jumps, ~4 orders of magnitude, from thousands to millions.

I will take a look and see if I can understand what's going on here, it may take some time for me to come back on this.

Reply:
A quick lookup tells me that function this is used extensively in `index` as well as with `Link`. This means that the slowness likely has little to do with wheels, and more to do with the fact that we're throwing literally every single URL we encounter through this as part of our lookup, which might not be operating with the maximal regard for avoiding duplicated work.

I understand the appeal of throwing a cache on this to make it faster -- and am not opposed to it -- but I do think that it's a good idea to spend a bit of time to actually profile the usage and call stacks before throwing on this cache here.

Reply:
> (If you have spent this time, let us know -- I might've missed information around this!)

Yes, I've been profiling this: https://github.com/pypa/pip/issues/12314#issuecomment-1749871193 (path_to_url is the far more light bluey green box). 

There are lots of other hot spots in this profile graph, but I just thought I'd start with the most simple looking offender.

Reply:
> Yes, I've been profiling this

The two big callers are `file_links()` and `page_candidates` in `_FlatDirectorySource`, and they do the same loop. So maybe put the result of that loop  in a cached property?

```python
class _FlatDirectorySource(LinkSource):
    def __init__(
        self,
        candidates_from_page: CandidatesFromPage,
        path: str,
    ) -> None:
        self._candidates_from_page = candidates_from_page
        self._path = pathlib.Path(os.path.realpath(path))
        self._file_urls = None

    @property
    def link(self) -> Optional[Link]:
        return None

    def _scan_dir(self):
        if self._file_urls is None:
            _file_urls = []
            for path in self._path.iterdir():
                url = path_to_url(str(path))
                _file_urls.append((url, _is_html_file(url)))
            self._file_urls = _file_urls

    def page_candidates(self) -> FoundCandidates:
        self._scan_dir()
        for url, html in self._file_urls:
            if html:
                yield from self._candidates_from_page(Link(url))

    def file_links(self) -> FoundLinks:
        self._scan_dir()
        return (Link(url) for url, html in self._file_urls if not html)
```

(Add type annotations as needed).

Reply:
I will try that locally and if successful create a new PR, unless you are wanting to given you have already wrote some code on this?

Reply:
No, go ahead. I haven't got the patience to work out the right type annotations :-)

Reply:
FYI, this approach doesn't work immediately because each call to `page_candidates` and `file_links` is from a seperate instance of `_FlatDirectorySource`. However there a number of possible solutions here, I will come up with one and submit a new PR.